### PR TITLE
Normalize ids before looking up in named export map

### LINKED ISSUE #177: namedExports is using non-normalized resolved module path as module id
On windows, this can conflict with plugins using other resolvers. For example typescript's built-in resolver returns module paths using forward slashes, while commonjs takes output of `resolve` at face value. Since that happens to return system native path with backslashes (in ver 1.1.7 at least), mishaps can result depending on which plugin happened to resolve a given module first. 

One way to fix it would be to normalize path before using it as a key.

### PR DESCRIPTION
**Pull request by bterlson**
This fixes an issue frequently hit when using the typescript rollup plugin. Because TypeScript always returns posix paths, but node resolve will normalize on a per-platform basis, comparisons must be carried out on normalized forms. This PR fixes some missing normalization and obviates rollup-plugin-typescript2's `rollupCommonJSResolveHack` option (/cc @ezolenko), and fixes #177 and probably a few others as well.

### CODE REVIEW

**Comment by bterlson** [comment]
Ok, I need a better way to test this on Linux. I guess paths on linux must always be posix paths else things break.

**Comment by shellscape** [comment]
I'd recommend stealing the CI config from rollup/rollup. We made some pretty solid containers that it uses.

File: src/index.js [hunk 1]
```diff
@@ -40,11 +40,14 @@ export default function commonjs(options = {}) {
 			} catch (err) {
 				resolvedId = resolve(id);
 			}
+
+			// ASSERT: resolvedId is a normalized path
```
Comment (lukastaegert) [inline src/index.js]: Not sure what to make of this comment and the one below. Are these TODOs?

File: test/test.js [hunk 2]
```diff
@@ -822,5 +822,41 @@ exports.shuffleArray = shuffleArray_1;
 `
 			);
 		});
+
+		it.only('normalizes paths used in the named export map', async () => {
```
Comment (lukastaegert) [inline test/test.js]: Probably not a good idea to check in an `.only` 😉

File: test/test.js [hunk 3]
```diff
@@ -822,5 +822,41 @@ exports.shuffleArray = shuffleArray_1;
...
+			function hookedResolve() {
+				const resolvePlugin = resolve();
+				const oldResolve = resolvePlugin.resolveId;
+				resolvePlugin.resolveId = async function(source) {
+					const result = await oldResolve.apply(resolvePlugin, arguments);
+					if (source === 'external') {
+						result.id = result.id.replace(/\\/g, '/');
+						return result;
+					} else if (source.match(/custom-named-exports/)) {
+						result.id = result.id.replace(/\//g, '\\');
```
Comment (lukastaegert) [inline test/test.js]: I would assume this does not work on Linux/Mac as a backslash here is treated as a legal file name character and will not be changed by `normalize`, at least to my knowledge. As this problem is only relevant on Windows, maybe it is possible to write this test in a way that it trivially passes on Linux? I.e. do not replace `/` with `\`, only the other way around? Or is the test meaningless in such a situation?

Comment (lukastaegert) [reply test/test.js]: Alternatively, you could also import `path` into the test and use `sep`.

**Comment by lukastaegert** [comment]
> Ok, I need a better way to test this on Linux. I guess paths on linux must always be posix paths else things break

Definitely, see my comment.

File: test/test.js (hunk 2 above)
Comment (bterlson) [reply test/test.js]: Pff, details.

File: test/test.js (hunk 3 above)
Comment (bterlson) [reply test/test.js]: I will push an update with the first approach later today (trivially passes on Linux).

File: src/index.js (hunk 1 above)
Comment (bterlson) [reply src/index.js]: Trying to point out that if the assert isn't true, bad things happen. Right now we can assume resolveId is normalized as both versions of `resolve` return normalized paths. If this changes in the future, things will break. I can just remove the comments.

**Comment by bterlson** [comment]
@lukastaegert the latter approach seemed best actually. I've also updated the comments to be more clear, see if you like them now 😀

File: src/index.js [hunk 4]
```diff
@@ -40,6 +40,10 @@ export default function commonjs(options = {}) {
 			} catch (err) {
 				resolvedId = resolve(id);
 			}
+
+			// Note: customNamedExport's keys must be normalized file paths.
+			// resolve and nodeResolveSync both return normalized file paths
+			// so no additional normalization is necessary.
```
Comment (lukastaegert) [inline src/index.js]: Yes, I'm cool with this, this is educational and will help people not familiar with the code 👍

**Comment by lukastaegert** [comment]
Would this be a breaking change or a bug fix?

**Comment by lukastaegert** [comment]
Thanks, then I'll make it a minor.

**Pull request merged**